pkg/controller/node/cidr_allocator.go

```diff
@@ -17,144 +17,169 @@ limitations under the License.
 package node
 
 import (
-	"encoding/binary"
 	"errors"
 	"fmt"
-	"math/big"
 	"net"
-	"sync"
+
+	"k8s.io/kubernetes/pkg/api"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/record"
+	"k8s.io/kubernetes/pkg/util/wait"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// controls how many NodeSpec updates NC can process in any moment.
+	cidrUpdateWorkers   = 10
+	cidrUpdateQueueSize = 5000
+	// podCIDRUpdateRetry controls the number of retries of writing Node.Spec.PodCIDR update.
+	podCIDRUpdateRetry = 5
 )
 
 var errCIDRRangeNoCIDRsRemaining = errors.New("CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range")
 
+type nodeAndCIDR struct {
+	cidr     *net.IPNet
+	nodeName string
+}
+
 // CIDRAllocator is an interface implemented by things that know how to allocate/occupy/recycle CIDR for nodes.
 type CIDRAllocator interface {
-	AllocateNext() (*net.IPNet, error)
-	Occupy(*net.IPNet) error
-	Release(*net.IPNet) error
+	AllocateOrOccupyCIDR(obj interface{}) error
+	ReleaseCIDR(obj interface{}) error
 }
 
 type rangeAllocator struct {
-	clusterCIDR     *net.IPNet
-	clusterIP       net.IP
-	clusterMaskSize int
-	subNetMaskSize  int
-	maxCIDRs        int
-	used            big.Int
-	lock            sync.Mutex
-	nextCandidate   int
+	cidrs                 *cidrSet
+	client                clientset.Interface
```

Comment:
nit: can you move client to the top? I would be more readable to have all cidr-related things together.